### PR TITLE
Account Transaction history is available. Closes #5

### LINKED ISSUE #5: Get All Account Transactions
User should be able to get a list of all Account Transactions for his Account.

### PR DESCRIPTION
**Pull request by SilasReinagel**

### CODE REVIEW

File: src/main/java/soolutions/hometaskapp/account/BasicAccount.java [hunk 1]
```diff
@@ -22,12 +30,22 @@ public boolean open() {
     }
 
     @Override
-    public Amount balance() {
-      return amount;
+    public void apply(Transaction transaction) {
+        if (!transaction.amount().matchesCurrency(currency))
```
Comment (vivekimsit) [inline src/main/java/soolutions/hometaskapp/account/BasicAccount.java]: I thinks we know too much about `Amount` object here.

File: src/main/java/soolutions/hometaskapp/account/BasicAccount.java [hunk 2]
```diff
@@ -22,12 +30,22 @@ public boolean open() {
...
+    @Override
+    public List<Transaction> transactions() {
+        return Collections.unmodifiableList(transactions);
     }
 
     @Override
-    public void deposit(Amount amount) {
-      this.amount = amount.add(this.amount);;
+    public Amount balance() {
+        Amount amount = new Amount(0, currency);
```
Comment (vivekimsit) [inline src/main/java/soolutions/hometaskapp/account/BasicAccount.java:43]: Like I said in my PR I am not a big fan of `new` can we use some factory? I am not sure but using a concrete class on right hand side is not a good idea.

File: src/main/java/soolutions/hometaskapp/account/BasicAccount.java (hunk 1 above)
Comment (SilasReinagel) [reply src/main/java/soolutions/hometaskapp/account/BasicAccount.java]: Any thought on how to change that?

Since Balance is now Lazy Evaluated, the previous unit test for adding EUR to a USD account wasn't throwing an exception. It makes business sense that one can't deposit EUR into a USD account.

Comment (llorllale) [reply src/main/java/soolutions/hometaskapp/account/BasicAccount.java]: @SilasReinagel let's trust our `Amount` object to be "smart" for now. Policies will be implemented later (you should've left a puzzle)

File: src/main/java/soolutions/hometaskapp/account/Account.java [hunk 3]
```diff
@@ -2,9 +2,12 @@
 
 import soolutions.hometaskapp.common.Amount;
 
+import java.util.List;
+
 public interface Account {
     boolean open();
-    void deposit(Amount amount);
+    void apply(Transaction transaction);
```
Comment (llorllale) [inline src/main/java/soolutions/hometaskapp/account/Account.java:9]: @SilasReinagel why was this renamed?

File: src/main/java/soolutions/hometaskapp/account/Account.java [hunk 4]
```diff
@@ -2,9 +2,12 @@
...
 
+    List<Transaction> transactions();
```
Comment (llorllale) [inline src/main/java/soolutions/hometaskapp/account/Account.java:11]: @SilasReinagel why `List`? Is there an ordering implicit here? Why not just `Collection`?

File: src/main/java/soolutions/hometaskapp/account/Deposit.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public Deposit(Amount amount, String description, Instant occurredAt) {
+        this.occurredAt = occurredAt;
+        this.description = description;
+        this.amount = amount;
+    }
+
+    @Override
+    public Instant occurredAt() {
```
Comment (llorllale) [inline src/main/java/soolutions/hometaskapp/account/Deposit.java]: @SilasReinagel have you read Yegor's take on "printers vs. getters"? Why do we have these getters? Do we need them right now?

File: src/test/java/soolutions/hometaskapp/account/AccountTests.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+public class AccountTests {
+    @Test
+    void openAccount() {
+        final Account account = createAccount();
+        assertEquals(true, account.open());
+        assertEquals(new Amount(0, "USD"), account.balance());
+    }
+
+    @Test
+    void depositAmount() {
```
Comment (llorllale) [inline src/test/java/soolutions/hometaskapp/account/AccountTests.java:30]: @SilasReinagel 1 assertion per test please

**Review by llorllale — CHANGES_REQUESTED** [review]
@SilasReinagel see my comments

File: src/main/java/soolutions/hometaskapp/account/Account.java (hunk 3 above)
Comment (SilasReinagel) [reply src/main/java/soolutions/hometaskapp/account/Account.java:9]: With the new interface `deposit` is too specific.

File: src/test/java/soolutions/hometaskapp/account/AccountTests.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        account.apply(new Deposit(new Amount(3, "USD"),"Sample Deposit 2"));
+
+        assertEquals(2, account.transactions().size());
+        assertEquals(new Amount(2, "USD"), account.transactions().get(0).amount());
+        assertEquals("Sample Deposit 1", account.transactions().get(0).description());
+        assertEquals(new Amount(3, "USD"), account.transactions().get(1).amount());
+        assertEquals("Sample Deposit 2", account.transactions().get(1).description());
+    }
+
+    private Account createAccount() {
```
Comment (vivekimsit) [inline src/test/java/soolutions/hometaskapp/account/AccountTests.java]: Let this method accept currency as a parameter too.

File: src/main/java/soolutions/hometaskapp/account/Deposit.java (hunk 5 above)
Comment (SilasReinagel) [reply src/main/java/soolutions/hometaskapp/account/Deposit.java]: Unit test won't pass without them. Do you have specific alternative way to make the test pass?

File: src/main/java/soolutions/hometaskapp/account/Account.java (hunk 4 above)
Comment (SilasReinagel) [reply src/main/java/soolutions/hometaskapp/account/Account.java:11]: Unit test won't compile without it. The transaction are currently in chronological order.

**Review by vivekimsit — APPROVED** [review]

File: src/main/java/soolutions/hometaskapp/account/BasicAccount.java (hunk 2 above)
Comment (SilasReinagel) [reply src/main/java/soolutions/hometaskapp/account/BasicAccount.java:43]: Partially agreed. 

On one hand, I think that `Amount` might be the wrong Value Type to be using in the system. 

It seems that if `Account` can be queried, he should be returning a primitive value. 

Also, I think that we should have a message that can be communicated through the system in the form of a data structure. Messages should be simple data structure with minimal behavior. 

Let's address those issues in separate tickets once we see the actual problems.

Comment (vivekimsit) [reply src/main/java/soolutions/hometaskapp/account/BasicAccount.java:43]: sure 👍

File: src/test/java/soolutions/hometaskapp/account/AccountTests.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        account.apply(new Deposit(new Amount(0, "USD"),"Sample Deposit 1"));
+        account.apply(new Deposit(new Amount(1, "USD"),"Sample Deposit 2"));
+
+        assertThrows(CurrencyMismatchException.class, () -> {
+            account.apply(new Deposit(new Amount(7, "EUR"),"Sample Deposit 3"));
+        }, "should fail");
+    }
+
+    @Test
+    void getAccountTransactions() {
```
Comment (llorllale) [inline src/test/java/soolutions/hometaskapp/account/AccountTests.java:60]: @SilasReinagel 1 assertion per test. Use [Matchers.contains(T...)](http://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/Matchers.html#contains(E...)) which combines checks on size and order.
